### PR TITLE
Adds FasterTableScanCrawler for crawling tables faster in python

### LINKED ISSUE #2190: [TECH DEBT]: Add log statements to `crawl_tables` to track listing databases and retriving table information
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

No log statements are visible

### Expected Behavior

Like the print statements in the attachment, but with a proper logger setup

[tables.scala.txt](https://github.com/user-attachments/files/16250585/tables.scala.txt)

### Steps To Reproduce

NA

### Cloud

Azure

### Operating System

macOS

### Version

latest via Databricks CLI

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by pritishpai**
## Changes
Adds logging to scala code 

### Linked issues

Resolves #2190 

### Functionality

- [x] added logging

### Tests

- [ ] manually tested
- [x] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

File: src/databricks/labs/ucx/hive_metastore/tables.scala [hunk 1]
```diff
@@ -10,114 +10,123 @@ import org.apache.spark.sql.catalyst.catalog.{CatalogTable, CatalogTableType}
 import org.apache.spark.sql.DataFrame
 import org.apache.spark.sql.functions.{col,lower,upper}
 import org.apache.spark.sql.catalyst.TableIdentifier
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.scala:12]: access scala classes via py4j gateway from python
```python
TableIdentifier = spark._jvm.org.apache.spark.sql.catalyst.TableIdentifier
```

**Comment by nfx** [comment]
@pritishpai use this as a starting point:

```
from functools import cached_property
from dataclasses import dataclass

import logging

logger = logging.getLogger('databricks')


@dataclass
class Table:
    catalog: str
    database: str
    name: str
    object_type: str
    table_format: str

    location: str | None = None
    view_text: str | None = None
    # really means migrated_to
    upgraded_to: str | None = None

    storage_properties: str | None = None
    is_partitioned: bool = False


class FasterTableScanHack:
    def __init__(self, spark: any):
        self._spark = spark

    @cached_property
    def _external_catalog(self):
        return self._spark._jsparkSession.sharedState().externalCatalog()

    def _iterator(self, result: any):
        it = result.iterator()
        while it.hasNext():
            yield it.next()

    def list_tables(self, database: str) -> list[str]:
        try:
            return list(self._iterator(self._external_catalog.listTables(database)))
        except Exception as err:
            logger.warning(f"Failed to list tables in {database}: {err}")

    def get_table(self, database: str, table: str) -> any:
        raw_table = self._external_catalog.getTable(database, table)
        table_format = raw_table.provider().getOrElse(None)
        if not table_format:
            table_format = "UNKNOWN"
        location_uri = raw_table.storage().locationUri().getOrElse(None)
        if location_uri:
            location_uri = location_uri.toString()
        view_text = raw_table.viewText()
        return Table(
            catalog='hive_metastore',
            database=database,
            name=table,
            object_type=raw_table.tableType().name(),
            table_format=table_format,
            location=location_uri,
            view_text=view_text,
        )

hack = FasterTableScanHack(spark)
hack.list_tables('default')

hack.get_table('default', 'ucx_oeto')
```

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 2]
```diff
@@ -391,19 +395,25 @@ def _try_load(self) -> Iterable[Table]:
...
+
     def _show_tables_in_database(self, catalog: str, database: str) -> list[Table]:
+
         table_rows: list[Table] = []
         try:
             logger.debug(f"[{catalog}.{database}] listing tables and views")
-            for row in self._fetch(
-                f"SHOW TABLES FROM {escape_sql_identifier(catalog)}.{escape_sql_identifier(database)}"
-            ):
+            external_catalog = self._spark._jsparkSession.sharedState().externalCatalog()
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.py:408]: the py4j hack has to be in a separate class. don't change TablesCrawler!

File: src/databricks/labs/ucx/hive_metastore/tables.scala [hunk 3]
```diff
@@ -20,85 +20,85 @@ case class TableError(catalog: String, database: String, name: String, error: St
 val failures = new ConcurrentLinkedQueue[TableError]()
 
 def metadataForAllTables(databases: Seq[String], queue: ConcurrentLinkedQueue[TableError]): DataFrame = {
-  import spark.implicits._
+    import spark.implicits._
 
-  val externalCatalog = spark.sharedState.externalCatalog
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.scala:25]: revert any changes to `tables.scala`.

File: src/databricks/labs/ucx/mixins/fixtures.py [hunk 4]
```diff
@@ -956,6 +956,174 @@ def sql_backend(ws, env_or_skip) -> StatementExecutionBackend:
     warehouse_id = env_or_skip("TEST_DEFAULT_WAREHOUSE_ID")
     return StatementExecutionBackend(ws, warehouse_id)
 
+# for testing, to be removed/modified with older fixture
+@pytest.fixture
```
Comment (nfx) [inline src/databricks/labs/ucx/mixins/fixtures.py:960]: you don't need to modify fixtures.py, revert any changes here.

File: tests/integration/hive_metastore/test_tables.py [hunk 5]
```diff
@@ -11,18 +12,20 @@
 
 
 @retried(on=[NotFound], timeout=timedelta(minutes=5))
-def test_describe_all_tables_in_databases(ws, sql_backend, inventory_schema, make_schema, make_table, env_or_skip):
+def test_describe_all_tables_in_databases(ws, sql_backend, spark_backend, inventory_schema, make_spark_schema, make_spark_table, env_or_skip):
```
Comment (nfx) [inline tests/integration/hive_metastore/test_tables.py:15]: don't change existing tests, add new ones.

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 6]
```diff
@@ -481,3 +481,75 @@ def _describe(self, catalog: str, database: str, table: str) -> Table | None:
...
+            name=table,
+            object_type=raw_table.tableType().name(),
+            table_format=table_format,
+            location=location_uri,
+            view_text=view_text,
+        )
+
+    def _all_databases(self) -> list[str]:
+        if not self._include_database:
+            return [row[0] for row in self._fetch("SHOW DATABASES")]
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.py:526]: there's no fetch method here. and use the previous optimized code to fetch databases. handle exceptions.

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 7]
```diff
@@ -481,3 +481,75 @@ def _describe(self, catalog: str, database: str, table: str) -> Table | None:
...
+            yield it.next()
+
+    def list_tables(self, database: str) -> list[str]:
+        try:
+            return list(self._iterator(self._external_catalog.listTables(database)))
+        except Exception as err:
+            logger.warning(f"Failed to list tables in {database}: {err}")
+
+    def get_table(self, database: str, table: str) -> any:
+        raw_table = self._external_catalog.getTable(database, table)
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.py:506]: handle exceptions

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 8]
```diff
@@ -481,3 +481,134 @@ def _describe(self, catalog: str, database: str, table: str) -> Table | None:
...
+
+                    redacted_key = "*******"
+
+                    if key == "personalAccessToken" or key.lower() == "password":
+                        value = redacted_key
+                    elif value is None:
+                        value = "None"
+
+                    formatted_items.append(f"{key}={value}")
+                storage_properties = f"[{', '.join(formatted_items)}]"
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.py:567]: refactor into a private method

File: tests/unit/hive_metastore/test_tables.py [hunk 9]
```diff
@@ -480,3 +482,106 @@ def test_in_place_migrate_hiveserde_sql_parsing_failure(caplog, ddl, expected_lo
...
+def test_fast_table_scan_crawler_already_crawled(caplog, mocker):
+    errors = {}
+    rows = {
+        "hive_metastore.inventory_database.tables": [
+            ("hive_metastore", "db1", "table1", "MANAGED", "DELTA", "dbfs:/location/table", None),
+            ("hive_metastore", "db1", "table2", "MANAGED", "DELTA", "/dbfs/location/table", None),
+            ("hive_metastore", "db1", "table3", "MANAGED", "DELTA", "dbfs:/mnt/location/table", None),
+        ],
+    }
+    backend = MockBackend(fails_on_first=errors, rows=rows)
```
Comment (nfx) [inline tests/unit/hive_metastore/test_tables.py:496]:
```suggestion
    sql_backend = MockBackend(fails_on_first=errors, rows=rows)
```

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 10]
```diff
@@ -481,3 +481,134 @@ def _describe(self, catalog: str, database: str, table: str) -> Table | None:
...
+        for database in databases:
+            logger.info(f"Scanning {database}")
+            table_names = [partial(self._list_tables, database)]
+            for table_batch in Threads.strict('listing tables', table_names):
+                if len(table_batch) > 0:
+                    for table in table_batch:
+                        tasks.append(partial(self._describe, catalog, database, table))
+        catalog_tables, errors = Threads.gather("describing tables in ", tasks)
+        if len(errors) > 0:
+            logger.error(f"Detected {len(errors)} errors while scanning tables in ")
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.py:613]:
```suggestion
            logger.warning(f"Detected {len(errors)} errors while scanning tables in ")
```
let's not fail the assessment if there are errors.

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 11]
```diff
@@ -481,3 +481,134 @@ def _describe(self, catalog: str, database: str, table: str) -> Table | None:
...
+                name=table,
+                object_type=raw_table.tableType().name(),
+                table_format=table_format,
+                location=location_uri,
+                view_text=view_text,
+                storage_properties=storage_properties,
+                is_partitioned=is_partitioned,
+            )
+        except Exception as e:  # pylint: disable=broad-exception-caught
+            logger.error(f"Couldn't fetch information for table {full_name} : {e}")
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.py:584]:
```suggestion
            logger.warning(f"Couldn't fetch information for table {full_name} : {e}")
```
error fails the job

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 12]
```diff
@@ -481,3 +481,134 @@ def _describe(self, catalog: str, database: str, table: str) -> Table | None:
...
+        if not self._include_database:
+            return list(self._iterator(self._external_catalog.listDatabases()))
+        return self._include_database
+
+    def _list_tables(self, database: str) -> list[str] | None:
+        try:
+            return list(self._iterator(self._external_catalog.listTables(database)))
+        except Exception as err:
+            logger.warning(f"Failed to list tables in {database}: {err}")
+            return None
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.py:530]:
```suggestion
    def _list_tables(self, database: str) -> list[str]:
        try:
            return list(self._iterator(self._external_catalog.listTables(database)))
        except Exception as err:
            logger.warning(f"Failed to list tables in {database}: {err}")
            return []
```
avoid `| None` whenever you can.

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 13]
```diff
@@ -481,3 +481,134 @@ def _describe(self, catalog: str, database: str, table: str) -> Table | None:
         except Exception as e:  # pylint: disable=broad-exception-caught
             logger.error(f"Couldn't fetch information for table {full_name} : {e}")
             return None
+
+
+class FasterTableScanCrawler(CrawlerBase):
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/tables.py:486]: add `use_experimental_tables_scan` to `WorkspaceConfig` and modify the `WorkflowsDeployment` to invoke that from if that config value is active. Add an integration test with that config active.

```
class Assessment(Workflow):
    ...
    @job_task(notebook="hive_metastore/tables.scala")
    def crawl_tables(self, ctx: RuntimeContext):
```

**Review by nfx — CHANGES_REQUESTED** [review]

**Comment by pritishpai** [comment]
Investigating why the test times out in the suite while it runs successfully in the IDE
============================= test session starts ==============================
collecting ... collected 1 item

unit/hive_metastore/test_tables.py::test_fast_table_scan_crawler_crawl_new PASSED [100%]

============================== 1 passed in 0.03s ===============================

File: src/databricks/labs/ucx/hive_metastore/tables.py (hunk 13 above)
Comment (pritishpai) [reply src/databricks/labs/ucx/hive_metastore/tables.py:486]: Will add this in the next PR to utilize the new FastTableScanCrawler class with integration tests

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**